File: mycollab-dao/src/main/java/com/esofthead/mycollab/core/persistence/service/ICrudService.java
/**
 * This file is part of mycollab-dao.
 *
 * mycollab-dao is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * mycollab-dao is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with mycollab-dao.  If not, see <http://www.gnu.org/licenses/>.
 */
package com.esofthead.mycollab.core.persistence.service;

import com.esofthead.mycollab.core.cache.CacheEvict;
import com.esofthead.mycollab.core.cache.CacheKey;
import com.esofthead.mycollab.core.cache.Cacheable;

import java.io.Serializable;
import java.util.List;

/**
 * @param <K>
 * @param <T>
 * @author MyCollab Ltd.
 * @since 1.0
 */
public interface ICrudService<K extends Serializable, T> extends IService {

    /**
     * @param record
     * @param username
     * @return
     */
    @CacheEvict
    Integer saveWithSession(@CacheKey T record, String username);

    /**
     * @param record
     * @param username
     * @return
     */
    @CacheEvict
    Integer updateWithSession(@CacheKey T record, String username);

    /**
     * @param record
     * @param username
     * @return
     */
    @CacheEvict
    Integer updateSelectiveWithSession(@CacheKey T record, String username);

    /**
     * @param record
     * @param primaryKeys
     * @param accountId
     */
    @CacheEvict
    void massUpdateWithSession(T record, List<K> primaryKeys, @CacheKey Integer accountId);

    /**
     * @param primaryKey
     * @param sAccountId
     * @return
     */
    @Cacheable
    T findByPrimaryKey(K primaryKey, @CacheKey Integer sAccountId);

    /**
     *
     * @param item
     * @param username
     * @param sAccountId
     */
    @CacheEvict
    void removeWithSession(T item, String username, @CacheKey Integer sAccountId);

    /**
     *
     * @param items
     * @param username
     * @param sAccountId
     */
    @CacheEvict
    void massRemoveWithSession(List<T> items, String username, @CacheKey Integer sAccountId);
}

File: mycollab-dao/src/main/java/com/mycollab/db/persistence/service/ICrudService.kt
/**
 * Copyright © MyCollab
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU Affero General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU Affero General Public License for more details.
 *
 * You should have received a copy of the GNU Affero General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>.
 */
package com.mycollab.db.persistence.service

import com.mycollab.core.cache.CacheEvict
import com.mycollab.core.cache.CacheKey
import com.mycollab.core.cache.Cacheable

import java.io.Serializable

/**
 * @param <K>
 * @param <T>
 * @author MyCollab Ltd.
 * @since 1.0
</T></K> */
interface ICrudService<K : Serializable, T> : IService {

    /**
     * @param record
     * @param username
     * @return
     */
    @CacheEvict
    fun saveWithSession(@CacheKey record: T, username: String?): Int

    /**
     * @param record
     * @param username
     * @return
     */
    @CacheEvict
    fun updateWithSession(@CacheKey record: T, username: String?): Int

    /**
     * @param record
     * @param username
     * @return
     */
    @CacheEvict
    fun updateSelectiveWithSession(@CacheKey record: T, username: String?): Int?

    /**
     * @param record
     * @param primaryKeys
     * @param accountId
     */
    @CacheEvict
    fun massUpdateWithSession(record: T, primaryKeys: List<K>, @CacheKey accountId: Int?)

    /**
     * @param primaryKey
     * @param sAccountId
     * @return
     */
    @Cacheable
    fun findByPrimaryKey(primaryKey: K, @CacheKey sAccountId: Int): T?

    /**
     * @param item
     * @param username
     * @param sAccountId
     */
    @CacheEvict
    fun removeWithSession(item: T, username: String?, @CacheKey sAccountId: Int)

    /**
     * @param items
     * @param username
     * @param sAccountId
     */
    @CacheEvict
    fun massRemoveWithSession(items: List<T>, username: String?, @CacheKey sAccountId: Int)
}
